crates/trie/sparse/src/trie.rs

```diff
@@ -0,0 +1,615 @@
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RevealedSparseTrie")
+            .field("nodes", &self.nodes)
+            .field("values", &self.values)
+            .field("prefix_set", &self.prefix_set)
+            .field("rlp_buf", &hex::encode(&self.rlp_buf))
+            .finish()
+    }
+}
+
+impl Default for RevealedSparseTrie {
+    fn default() -> Self {
+        Self {
+            nodes: HashMap::from_iter([(Nibbles::default(), SparseNode::Empty)]),
+            values: HashMap::default(),
+            prefix_set: PrefixSetMut::default(),
+            rlp_buf: Vec::new(),
+        }
+    }
+}
+
+impl RevealedSparseTrie {
+    /// Create new revealed sparse trie from the given root node.
+    pub fn from_root(node: TrieNode) -> alloy_rlp::Result<Self> {
+        let mut this = Self {
+            nodes: HashMap::default(),
+            values: HashMap::default(),
+            prefix_set: PrefixSetMut::default(),
+            rlp_buf: Vec::new(),
+        };
+        this.reveal_node(Nibbles::default(), node)?;
+        Ok(this)
+    }
+
+    /// Reveal the trie node only if it was not known already.
+    pub fn reveal_node(&mut self, path: Nibbles, node: TrieNode) -> alloy_rlp::Result<()> {
+        // TODO: revise all inserts to not overwrite existing entries
+        match node {
+            TrieNode::EmptyRoot => {
+                debug_assert!(path.is_empty());
+                self.nodes.insert(path, SparseNode::Empty);
+            }
+            TrieNode::Branch(branch) => {
+                let mut stack_ptr = branch.as_ref().first_child_index();
+                for idx in CHILD_INDEX_RANGE {
+                    if branch.state_mask.is_bit_set(idx) {
+                        let mut child_path = path.clone();
+                        child_path.push_unchecked(idx);
+                        self.reveal_node_or_hash(child_path, &branch.stack[stack_ptr])?;
+                        stack_ptr += 1;
+                    }
+                }
+                self.nodes
+                    .insert(path, SparseNode::Branch { state_mask: branch.state_mask, hash: None });
+            }
+            TrieNode::Extension(ext) => {
+                let mut child_path = path.clone();
+                child_path.extend_from_slice_unchecked(&ext.key);
+                self.reveal_node_or_hash(child_path, &ext.child)?;
+                self.nodes.insert(path, SparseNode::Extension { key: ext.key, hash: None });
+            }
+            TrieNode::Leaf(leaf) => {
+                let mut full = path.clone();
+                full.extend_from_slice_unchecked(&leaf.key);
+                self.values.insert(full, leaf.value);
+                self.nodes.insert(path, SparseNode::new_leaf(leaf.key));
+            }
+        }
+
+        Ok(())
+    }
+
+    fn reveal_node_or_hash(&mut self, path: Nibbles, child: &[u8]) -> alloy_rlp::Result<()> {
+        if child.len() == B256::len_bytes() + 1 {
```

Comment:
can we add a comment here that if the length is 33 bytes, then it's an RLP-encoded string, and we need to skip the first byte that is the length of the string